devito/finite_differences/differentiable.py

```diff
@@ -186,20 +187,93 @@ def _has(self, pattern):
         return super(Differentiable, self)._has(pattern)
 
 
-class Add(sympy.Add, Differentiable):
-    pass
+class DifferentiableOp(Differentiable):
 
+    def __new__(cls, *args, **kwargs):
+        obj = cls.__base__.__new__(cls, *args, **kwargs)
 
-class Mul(sympy.Mul, Differentiable):
-    pass
+        # Unfortunately SymPy may build new sympy.core objects (e.g., sympy.Add),
+        # so here we have to rebuild them as devito.core objects
+        obj = diffify(obj)
```

Comment:
I thought you didn't like that.......

Reply:
this is the point of this PR -- I'm going back to rebuilding in place of monkey-patching. I thought deeply about how to do this, including a few other ways, and with all that in mind I think I can say this is the way to go. It's not a good one, but you know...we do what we can do. 

Unless you're talking about the name... I don't like it, but it makes a parallel with sympy